File: Blackjack21-JAVA/Blackjack21Library/src/Blackjack21/Game/Logic/Player.java
package Blackjack21.Game.Logic;

import Blackjack21.Game.Exceptions.InvalidCardActionException;
import Blackjack21.Game.Exceptions.InvalidHandSplitException;
import Blackjack21.Game.Exceptions.PlayerHandNotSplitException;
import Blackjack21.Game.Model.Card;
import Blackjack21.Game.Model.PlayerHandResult;
import Blackjack21.Game.Model.PlayerHandType;

/**
 * A blackjack player object
 */
public class Player
{

    // Store our hand in a list, a player can have an additional hand if they choose to split.
    private final PlayerHand[] _playerHand;  // a player can only split his hand once there for we use an array

    private PlayerHandType _playerHandType;
    private boolean _playerInsured;
    private boolean _playerIsuranceCorrect;
    private final String _playerName;

    /**
     * initializes the player object
     *
     * @param playerName The player's name
     */
    public Player(String playerName)
    {
        // set player default hands to 2
        this._playerHand = new PlayerHand[]
        {
            new PlayerHand(), new PlayerHand()
        };
        this._playerName = playerName;
        this._playerHandType = PlayerHandType.SINGLE_HAND;
        this._playerIsuranceCorrect = false;
        this._playerInsured = false;
    }

    /**
     * Returns a value to indicate if the player has a single or split hand
     *
     * @return
     */
    public PlayerHandType getCurrentHandType()
    {
        return this._playerHandType;
    }

    /**
     * Gets the players first hand
     *
     * @return
     */
    public PlayerHand getFirstHand()
    {
        return this._playerHand[0];
    }

    /**
     * Gets the player's split hand, throws an exception if the player does not
     * have a split hand
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public PlayerHand getSplitHand() throws PlayerHandNotSplitException
    {
        if (this._playerHandType != PlayerHandType.SPLIT_HAND)
        {
            throw new PlayerHandNotSplitException();
        }

        return this._playerHand[1];
    }

    public void setIsurance(int dealerHandValue)
    {
        this._playerInsured = true;
        this._playerIsuranceCorrect = dealerHandValue == PlayerHand.MAX_HAND_VALUE;
    }

    /**
     * Gets if the player is insured
     *
     * @return
     */
    public boolean getPlayerInsured()
    {
        return _playerInsured;
    }

    /**
     * Gets if the players insurance was correct
     *
     * @return
     */
    public boolean getPlayerIsuranceCorrect()
    {
        return this._playerIsuranceCorrect;
    }

    /**
     * Returns the player's name, throws an exception if the player name is
     * empty
     *
     * @return
     */
    public String getPlayerName()
    {
        return _playerName;
    }

    /**
     * Gets of the player's hand can be split
     *
     * @return
     */
    public boolean getCanSplitHand()
    {
        return this._playerHandType == PlayerHandType.SINGLE_HAND && this.getFirstHand().getHasTwoPairs();
    }

    /**
     * Gets if the player can double their hand
     *
     * @return
     */
    public boolean getCanDouble()
    {
        return this.getFirstHand().getCardCount() == 2 && this.getFirstHandValue() >= 9 && this.getFirstHandValue() <= 11;
    }

    /**
     * Gets if the player can add more cards to their hand
     *
     * @return
     */
    public boolean getCanHit()
    {
        return this.getFirstHandValue() < PlayerHand.MAX_HAND_VALUE;
    }

    /**
     * Gets if the player can stand
     *
     * @return
     */
    public boolean getCanStand()
    {
        return this.getFirstHandValue() < PlayerHand.MAX_HAND_VALUE;
    }

    /**
     * Gets if the player can fold their hand
     *
     * @return
     */
    public boolean getCanFold()
    {
        return this.getCurrentHandType() == PlayerHandType.SINGLE_HAND && this.getFirstHand().getCardCount() == 2;
    }

    /**
     * Double the players first hand only
     *
     * @param card Card to add to the doubled hand
     * @throws Blackjack21.Game.Exceptions.InvalidCardActionException
     */
    public void doubleHand(Card card) throws InvalidCardActionException
    {
        if (!getCanDouble())
        {
            throw new InvalidCardActionException("Cannot double the player's hand, conditions not met");
        }

        getFirstHand().doubleHandCard(card);
    }

    /**
     * Fold the player's hand
     *
     * @throws Blackjack21.Game.Exceptions.InvalidCardActionException
     */
    public void fold() throws InvalidCardActionException
    {
        if (!getCanFold())
        {
            throw new InvalidCardActionException("Cannot fold the player's hand, conditions not met");
        }

        this.getFirstHand().fold();
    }

    /**
     * Stand on the player's current hand
     *
     * @throws Blackjack21.Game.Exceptions.InvalidCardActionException
     */
    public void stand() throws InvalidCardActionException
    {
        if (!getCanStand())
        {
            throw new InvalidCardActionException("Cannot stand the player's hand, conditions not met");
        }

        this.getFirstHand().stand();
    }

    /**
     * Gets if the player can double their hand
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public boolean getCanDoubleSplitHand() throws PlayerHandNotSplitException
    {
        return this.getSplitHand().getCardCount() == 2 && this.getSplitHandValue() >= 9 && this.getSplitHandValue() <= 11;
    }

    /**
     * Gets if the player can add more cards to their split hand
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public boolean getCanHitSplitHand() throws PlayerHandNotSplitException
    {
        return this.getSplitHandValue() < PlayerHand.MAX_HAND_VALUE;
    }

    /**
     * Gets if the player can stand on their split hand
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public boolean getCanStandSplitHand() throws PlayerHandNotSplitException
    {
        return this.getSplitHandValue() < PlayerHand.MAX_HAND_VALUE;
    }

    /**
     * Double the player's split hand only
     *
     * @param card Card to add to the doubled hand
     * @throws Blackjack21.Game.Exceptions.InvalidCardActionException
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void doubleSplitHand(Card card) throws InvalidCardActionException, PlayerHandNotSplitException
    {
        if (!getCanDoubleSplitHand())
        {
            throw new InvalidCardActionException("Cannot double the player's hand, conditions not met");
        }

        this.getSplitHand().doubleHandCard(card);
    }

    /**
     * Stand on the player's current hand
     *
     * @throws Blackjack21.Game.Exceptions.InvalidCardActionException
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void standSplitHand() throws InvalidCardActionException, PlayerHandNotSplitException
    {
        if (!getCanStandSplitHand())
        {
            throw new InvalidCardActionException("Cannot stand the player's split hand, conditions not met");
        }

        this.getSplitHand().stand();
    }

    /**
     * Splits the players hand into two, throws an exception if the hand does
     * not meet split conditions
     *
     * @throws Blackjack21.Game.Exceptions.InvalidHandSplitException
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void splitHandCards() throws InvalidHandSplitException, PlayerHandNotSplitException
    {
        if (this._playerHandType == PlayerHandType.SINGLE_HAND && this.getFirstHand().getHasTwoPairs() == true)
        {
            this._playerHandType = PlayerHandType.SPLIT_HAND;
            addSplitHandCard(this._playerHand[0].splitHandCard());
            return;
        }

        throw new InvalidHandSplitException();
    }

    /**
     * Adds a card to the player's first hand
     *
     * @param card Card to add to the player's hand
     */
    public void addFirstHandCard(Card card)
    {
        this._playerHand[0].addCard(card);
    }

    /**
     * Adds cards to the player's first hand
     *
     * @param cards Card to add to the player's hand
     */
    public void addFirstHandCards(Card[] cards)
    {
        for (Object obj : cards)
        {
            Card card = (Card) obj;
            addFirstHandCard(card);
        }
    }

    /**
     * Adds a card to the players split hand, provided the player has a split
     * hand, throws an exception if the player's hand is not split
     *
     * @param card Card to add to the player's split hand
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void addSplitHandCard(Card card) throws PlayerHandNotSplitException
    {
        if (this._playerHandType != PlayerHandType.SPLIT_HAND)
        {
            throw new PlayerHandNotSplitException();
        }

        this._playerHand[1].addCard(card);
    }

    /**
     * Adds cards to the players split hand, provided the player has a split
     * hand
     *
     * @param cards Cards to add to the player's split hand
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void addSplitHandCards(Card[] cards) throws PlayerHandNotSplitException
    {
        for (Object obj : cards)
        {
            Card card = (Card) obj;
            addSplitHandCard(card);
        }
    }

    /**
     * Get the value of the player's first hand
     *
     * @return
     */
    public int getFirstHandValue()
    {
        return this._playerHand[0].getHandValue();
    }

    /**
     * Gets the player's first hand result
     *
     * @return
     */
    public PlayerHandResult getFirstHandResult()
    {
        return this._playerHand[0].getHandResult();
    }

    /**
     * Get the value of the player's split hand, throws an exception if the
     * player does not have a split hand
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public int getSplitHandValue() throws PlayerHandNotSplitException
    {
        if (this._playerHandType != PlayerHandType.SPLIT_HAND)
        {
            throw new PlayerHandNotSplitException();
        }

        return this._playerHand[1].getHandValue();
    }

    /**
     * Gets the player's split hand result
     *
     * @return
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public PlayerHandResult getSplitHandResult() throws PlayerHandNotSplitException
    {
        if (this._playerHandType != PlayerHandType.SPLIT_HAND)
        {
            throw new PlayerHandNotSplitException();
        }

        return this._playerHand[1].getHandResult();
    }

    /**
     * Sets the players hand result based on the dealers hand
     *
     * @param dealer The dealers object
     * @throws Blackjack21.Game.Exceptions.PlayerHandNotSplitException
     */
    public void setHandResult(Player dealer) throws PlayerHandNotSplitException
    {
        if (this.getCurrentHandType() == PlayerHandType.SINGLE_HAND)
        {
            this.getFirstHand().setHandResult(dealer.getFirstHand(), true);
        }
        else
        {
            this.getFirstHand().setHandResult(dealer.getFirstHand(), false);
            this.getSplitHand().setHandResult(dealer.getFirstHand(), false);
        }
    }

    /**
     * Clear the player's active hand cards
     */
    public void clearHands()
    {
        this._playerHand[0].clearCards();
        this._playerHand[1].clearCards();
        // clear split hand cards
        this._playerHandType = PlayerHandType.SINGLE_HAND;
        this._playerIsuranceCorrect = false;
        this._playerInsured = false;
    }

    /**
     * Returns the player object as a string
     *
     * @return return the player as a string
     */
    @Override
    public String toString()
    {

        String result;
        result = "Player: " + getPlayerName();
        if (this.getPlayerInsured())
        {
            result += "\n\n-- Player placed an insurance bet and ".toUpperCase() + (this.getPlayerIsuranceCorrect() ? "WON" : "LOST") + "";
        }

        result += "\n\n-- Hand Cards: \n";
        result += getFirstHand().toString();
        if (this._playerHandType == PlayerHandType.SPLIT_HAND)
        {
            result += "\n-- Split Hand Cards: ";
            result += getFirstHand().toString();
        }

        result += "------------------------------\n";
        return result;

    }

}

File: BlackJack-CSHARP/Blackjack21.Game/Game/Logic/Player.cs
using System.Collections.Generic;
using Blackjack21.Game.Model;
using Blackjack21.Game.Exceptions;

namespace Blackjack21.Game.Logic
{
    /// <summary>
    /// A blackjack player object
    /// </summary>
    public class Player
    {
        // Store our hand in a list, a player can have an additional hand if they choose to split.
        private readonly List<PlayerHand> _playerHand; // a player can only split his hand once there for we use an array
        private PlayerHandType _playerHandType;
        private bool _playerInsured;
        private bool _playerIsuranceCorrect;

        private readonly string _playerName;

        /// <summary>
        /// initializes the player objecy
        /// </summary>
        /// <param name="playerName">The player's name</param>
        public Player(string playerName = "Player")
        {
            // set player default hands to 2
            this._playerHand = new List<PlayerHand>();
            this._playerName = playerName;
            /*this._playerHandType = PlayerHandType.SINGLE_HAND;
            this._playerIsuranceCorrect = false;
            this._playerInsured = false;*/

            ClearHands();
        }

        /// <summary>
        /// Returns a value to indicate if the player has a single or split hand
        /// </summary>
        public PlayerHandType CurrentHandType
        {
            get
            {
                return this._playerHandType;
            }
        }

        /// <summary>
        /// Gets the players first hand
        /// </summary>
        public PlayerHand FirstHand
        {
            get
            {
                return this._playerHand[0];
            }
        }

        /// <summary>
        /// Gets the player's split hand, throws an exception if the player does not have a split hand
        /// </summary>
        public PlayerHand SplitHand
        {
            get
            {
                if (this._playerHandType != PlayerHandType.SPLIT_HAND)
                {
                    throw new PlayerHandNotSplitException();
                }
                return this._playerHand[1];
            }
        }

        public void SetIsurance(int dealerHandValue)
        {
            this._playerInsured = true;
            this._playerIsuranceCorrect = dealerHandValue == PlayerHand.MAX_HAND_VALUE;
        }

        /// <summary>
        /// Gets if the player is insured
        /// </summary>
        public bool PlayerInsured
        {
            get
            {
                return _playerInsured;
            }
        }

        /// <summary>
        /// Gets if the players insurance was correct
        /// </summary>
        public bool PlayerIsuranceCorrect
        {
            get
            {
                return this._playerIsuranceCorrect;
            }
        }


        /// <summary>
        /// Returns the player's name, throws an exception if the player name is empty
        /// </summary>
        public string PlayerName
        {
            get
            {
                return _playerName;
            }
        }


        /// <summary>
        /// Gets of the player's hand can be split
        /// </summary>
        public bool CanSplitHand
        {
            get
            {
                return this._playerHandType == PlayerHandType.SINGLE_HAND && this.FirstHand.HasTwoPairs;
            }
        }


        /// <summary>
        /// Gets if the player can double their hand
        /// </summary>
        public bool CanDouble
        {
            get
            {
                return this.FirstHand.CardCount == 2 && this.FirstHandValue >= 9 && this.FirstHandValue <= 11;
            }
        }

        /// <summary>
        /// Gets if the player can add more cards to their hand
        /// </summary>
        public bool CanHit
        {
            get
            {
                return this.FirstHandValue < PlayerHand.MAX_HAND_VALUE;
            }
        }

        /// <summary>
        /// Gets if the player can stand
        /// </summary>
        public bool CanStand
        {
            get
            {
                return this.FirstHandValue < PlayerHand.MAX_HAND_VALUE;
            }
        }

        /// <summary>
        /// Gets if the player can fold their hand
        /// </summary>
        public bool CanFold
        {
            get
            {
                return this.CurrentHandType == PlayerHandType.SINGLE_HAND && this.FirstHand.CardCount == 2;
            }
        }


        /// <summary>
        /// Double the players first hand only
        /// </summary>
        /// <param name="card">Card to add to the doubled hand</param>
        public void DoubleHand(Card card)
        {
            if (!CanDouble)
            {
                throw new InvalidCardActionException("Cannot double the player's hand, conditions not met");
            }
            FirstHand.DoubleHandCard(card);
        }

        /// <summary>
        /// Fold the player's hand
        /// </summary>
        public void Fold()
        {
            if (!CanFold)
            {
                throw new InvalidCardActionException("Cannot fold the player's hand, conditions not met");
            }
            this.FirstHand.Fold();
        }

        /// <summary>
        /// Stand on the player's current hand
        /// </summary>
        public void Stand()
        {
            if (!CanStand)
            {
                throw new InvalidCardActionException("Cannot stand the player's hand, conditions not met");
            }
            this.FirstHand.Stand();
        }



        /// <summary>
        /// Gets if the player can double their hand
        /// </summary>
        public bool CanDoubleSplitHand
        {
            get
            {
                return this.SplitHand.CardCount == 2 && this.SplitHandValue >= 9 && this.SplitHandValue <= 11;
            }
        }

        /// <summary>
        /// Gets if the player can add more cards to their split hand
        /// </summary>
        public bool CanHitSplitHand
        {
            get
            {
                return this.SplitHandValue < PlayerHand.MAX_HAND_VALUE;
            }
        }

        /// <summary>
        /// Gets if the player can stand on their split hand
        /// </summary>
        public bool CanStandSplitHand
        {
            get
            {
                return this.SplitHandValue < PlayerHand.MAX_HAND_VALUE;
            }
        }


        /// <summary>
        /// Double the player's split hand only
        /// </summary>
        /// <param name="card">Card to add to the doubled hand</param>
        public void DoubleSplitHand(Card card)
        {
            if (!CanDoubleSplitHand)
            {
                throw new InvalidCardActionException("Cannot double the player's hand, conditions not met");
            }
            this.SplitHand.DoubleHandCard(card);
        }



        /// <summary>
        /// Stand on the player's current hand
        /// </summary>
        public void StandSplitHand()
        {
            if (!CanStandSplitHand)
            {
                throw new InvalidCardActionException("Cannot stand the player's split hand, conditions not met");
            }
            this.SplitHand.Stand();
        }



        /// <summary>
        /// Splits the players hand into two, throws an exception if the hand does not meet split conditions
        /// </summary>
        public void SplitHandCards()
        {
            if (this._playerHandType == PlayerHandType.SINGLE_HAND && this.FirstHand.HasTwoPairs)
            {
                this._playerHandType = PlayerHandType.SPLIT_HAND;
                AddSplitHandCard(this._playerHand[0].SplitHandCard());
                return;
            }
            throw new InvalidHandSplitException();
        }

        /// <summary>
        /// Adds a card to the player's first hand
        /// </summary>
        /// <param name="card">Card to add to the player's hand</param>
        public void AddFirstHandCard(Card card)
        {
            this._playerHand[0].AddCard(card);
        }

        /// <summary>
        ///  Adds cards to the player's first hand
        /// </summary>
        /// <param name="cards">Card to add to the player's hand</param>
        public void AddFirstHandCards(Card[] cards)
        {
            foreach (Card card in cards)
            {
                AddFirstHandCard(card);
            }
        }


        /// <summary>
        ///  Adds a card to the players split hand, provided the player has a split hand, throws an exception if the player's hand is not split
        /// </summary>
        /// <param name="card">Card to add to the player's split hand</param>
        public void AddSplitHandCard(Card card)
        {
            if (this._playerHandType != PlayerHandType.SPLIT_HAND)
            {
                throw new PlayerHandNotSplitException();
            }

            this._playerHand.Add(new PlayerHand());
            this._playerHand[1].AddCard(card);
        }

        /// <summary>
        /// Adds cards to the players split hand, provided the player has a split hand
        /// </summary>
        /// <param name="cards">Cards to add to the player's split hand</param>
        public void AddSplitHandCards(Card[] cards)
        {
            foreach (Card card in cards)
            {
                AddSplitHandCard(card);
            }
        }

        /// <summary>
        /// Get the value of the player's first hand
        /// </summary>
        public int FirstHandValue
        {
            get
            {
                return this._playerHand[0].HandValue;
            }
        }

        /// <summary>
        /// Gets the player's first hand result
        /// </summary>
        public PlayerHandResult FirstHandResult
        {
            get
            {
                return this._playerHand[0].HandResult;
            }
        }

        /// <summary>
        /// Get the value of the player's split hand, throws an exception if the player does not have a split hand 
        /// </summary>
        public int SplitHandValue
        {
            get
            {
                if (this._playerHandType != PlayerHandType.SPLIT_HAND)
                {
                    throw new PlayerHandNotSplitException();
                }
                return this._playerHand[1].HandValue;
            }
        }

        /// <summary>
        /// Gets the player's split hand result
        /// </summary>
        public PlayerHandResult SplitHandResult
        {
            get
            {
                if (this._playerHandType != PlayerHandType.SPLIT_HAND)
                {
                    throw new PlayerHandNotSplitException();
                }
                return this._playerHand[1].HandResult;
            }
        }

        /// <summary>
        /// Sets the players hand result based on the dealers hand
        /// </summary>
        /// <param name="dealer">The dealers object</param>
        public void SetHandResult(Player dealer)
        {

            if (this.CurrentHandType == PlayerHandType.SINGLE_HAND)
            {
                this.FirstHand.SetPlayerHandResult(dealer.FirstHand, true);
            }
            else
            {
                this.FirstHand.SetPlayerHandResult(dealer.FirstHand);
                this.SplitHand.SetPlayerHandResult(dealer.FirstHand);
            }

        }

        /// <summary>
        /// Clear the player's active hand cards
        /// </summary>
        public void ClearHands()
        {

            foreach (PlayerHand hand in this._playerHand)
            {
                hand.ClearCards();
            }

            _playerHand.Clear();

            // Add a new player hand
            _playerHand.Add(new PlayerHand());

            this._playerHandType = PlayerHandType.SINGLE_HAND;
            this._playerIsuranceCorrect = false;
            this._playerInsured = false;
        }

        /// <summary>
        /// Returns the player object as a string
        /// </summary>
        /// <returns>retuns the player as a string</returns>
        public override string ToString()
        {
            string result;

            result = "Player: " + PlayerName;

            if (this.PlayerInsured)
            {
                result += "\n\n-- Player placed an insurance bet and ".ToUpper() + (this.PlayerIsuranceCorrect ? "WON" : "LOST") + "";
            }

            result += "\n\n-- Hand Cards: \n";
            result += FirstHand.ToString();


            if (this._playerHandType == PlayerHandType.SPLIT_HAND)
            {
                result += "\n-- Split Hand(s) Cards: \n";

                for (int x = 1; x < this._playerHand.Count; x++)
                {
                    PlayerHand hand = this._playerHand[x];
                    result += "** \n";
                    result += hand.ToString();
                    result += "\n";
                }
              
            }
            result += "------------------------------\n";

            return result;
        }

    }
}
